google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/OpenTelemetryUtil.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    return attributesBuilder;
+  }
+
+  /**
+   * Creates, starts, and returns a publish RPC span for the given message batch. Bi-directional
+   * links with the publisher parent span are created for sampled messages in the batch.
+   */
+  public static final Span startPublishRpcSpan(
+      Tracer tracer,
+      String topic,
+      List<PubsubMessageWrapper> messages,
+      boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && tracer != null) {
+      TopicName topicName = TopicName.parse(topic);
+      Attributes attributes =
+          createCommonSpanAttributesBuilder(
+                  topicName.getTopic(), topicName.getProject(), "Publisher.publishCall", "publish")
+              .put(SemanticAttributes.MESSAGING_BATCH_MESSAGE_COUNT, messages.size())
+              .build();
+      Span publishRpcSpan =
+          tracer
+              .spanBuilder(topicName.getTopic() + PUBLISH_RPC_SPAN_SUFFIX)
+              .setSpanKind(SpanKind.CLIENT)
+              .setAllAttributes(attributes)
+              .startSpan();
+
+      for (PubsubMessageWrapper message : messages) {
+        if (message.getPublisherSpan().getSpanContext().isSampled()) {
+          Attributes linkAttributes =
+              Attributes.builder().put(SemanticAttributes.MESSAGING_OPERATION, "publish").build();
+          publishRpcSpan.addLink(message.getPublisherSpan().getSpanContext(), linkAttributes);
+          message.getPublisherSpan().addLink(publishRpcSpan.getSpanContext(), linkAttributes);
+          message.addPublishStartEvent();
+        }
+      }
+      return publishRpcSpan;
+    }
+    return null;
+  }
+
+  /** Ends the given publish RPC span if it exists. */
+  public static final void endPublishRpcSpan(
+      Span publishRpcSpan, boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && publishRpcSpan != null) {
+      publishRpcSpan.end();
+    }
+  }
+
+  /**
+   * Sets an error status and records an exception when an exception is thrown when publishing the
+   * message batch.
+   */
+  public static final void setPublishRpcSpanException(
+      Span publishRpcSpan, Throwable t, boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && publishRpcSpan != null) {
+      publishRpcSpan.setStatus(StatusCode.ERROR, "Exception thrown on publish RPC.");
+      publishRpcSpan.recordException(t);
+      publishRpcSpan.end();
+    }
+  }
+
+  /**
+   * Creates, starts, and returns spans for ModAck, Nack, and Ack RPC requests. Bi-directional links
+   * to parent subscribe span for sampled messages are added.
+   */
+  public static final Span startSubscribeRpcSpan(
+      Tracer tracer,
+      String subscription,
+      String rpcOperation,
+      List<PubsubMessageWrapper> messages,
+      int ackDeadline,
+      boolean isReceiptModack,
+      boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && tracer != null) {
```

Comment:
It looks like we are passing `enableOpenTelemetryTracing` and `tracer` all the time, it seems we can at least extract them as fields and make `OpenTelemetryUtil` as a context aware wrapper of `Tracer`. In addition, we can have create an interface `PubsubTracer` that have all the injection points, and we can create an `OpenTelemetryPubsubTracer` only if `enableOpenTelemetryTracing` is true. Similar concepts can be applied to `PubsubMessageWrapper` as well. 

We have this architecture in gax. There is a [ApiTracer](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/tracing/ApiTracer.java) that defines all the injection points, and there is an [OpenCensusTracer](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java) that is created conditionally, so we don't have to always check if instrumentation is enabled or not. This also has a clearer separation of concern that in case we need to move away from OpenTelemetry, we don't have to write all the code, only the implementation of the interface.

Reply:
In the [latest commit](https://github.com/googleapis/java-pubsub/pull/2086/commits/d257b6bb59ecbc607b1333580d615cff79898c85), I've refactored the OpenTelemetry implementation to use this architecture. Everything is controlled through a `PubsubTracer` object that performs no-ops unless OpenTelemetry is enabled.